api/controllers/emailSubscribe.js

```diff
@@ -59,177 +59,231 @@ exports.publicHead = async function (args, res) {
+      defaultLog.info('Sent confirmation email for new subscription', { email: requestedEmail, projects: projectNames });
+
+      return sendGenericSuccess();
     }
-  });
 
-  // check if already exists
-  // if so either update with the new project or exit gracefully
-  await EmailSubscribe.findOne({ _schemaName: 'EmailSubscribe', email: emailSubscribe.email }, null, async function (err, entity) {
-    if (entity) {
-      existingEmailId = entity._id;
-      existingProjectArray = entity.project;
-      alreadyConfirmed = entity.confirmed;
-      confirmKey = entity.confirmKey;
-      // check if pushed project is in array
-      if (existingProjectArray.includes(mongoose.Types.ObjectId(obj.project)) ) {
-        isDuplicate = true;
-      }
+    const alreadySubscribedToProject = subscription.project.some(
+      projId => projId && projId.toString() === requestedProjectId.toString()
+    );
+
+    let subscriptionChanged = false;
+
+    if (!alreadySubscribedToProject) {
+      subscription.project.push(requestedProjectId);
+      subscriptionChanged = true;
     }
 
-    if (existingEmailId && isDuplicate ) {
-      // Project and email already exists so exit gracefully
-      defaultLog.info('User has already signed up for the project', existingEmailId);
-      return Actions.sendResponse(res, 200, '200');
-    } else if (existingEmailId) {
-      // New project for an existing email
-      existingProjectArray.push(mongoose.Types.ObjectId(obj.project));
-      await EmailSubscribe.updateOne({ _id: existingEmailId }, { $set: { project: existingProjectArray } }, async function (err, entity) {
-        if (err) throw new Error(err);
-        if (entity) {
-          Utils.recordAction('Put', 'EmailSubscribe', 'public', existingEmailId);
-          defaultLog.info('New project added to email subscribe:', entity._id);
-        }
-      });
-      // have they already confirmed their email?
-      // if so, send the welcome for the new project
-      if (alreadyConfirmed) {
-        defaultLog.info('Email was already confirmed - sending welcome email for project/email', projectName, emailSubscribe.email);
-        await Email.sendWelcomeEmail(projectName, emailSubscribe.email)
-          .then(() => undefined)
-          .catch((e) => {
-            defaultLog.error('Error sending welcome email:', e);
-            return Actions.sendResponse(res, 500, e);
-          });
+    // Handle unconfirmed subscriptions
+    if (!subscription.confirmed) {
+      const TWENTY_FOUR_HOURS_MS = 24 * 60 * 60 * 1000;
+      const twentyFourHoursAgo = new Date(now.getTime() - TWENTY_FOUR_HOURS_MS);
+      const shouldSendConfirmEmail = !subscription.dateSubscribed || subscription.dateSubscribed <= twentyFourHoursAgo;
+
+      if (shouldSendConfirmEmail) {
+        const projectNames = await fetchProjectNames(Project, subscription.project, requestedProjectName);
+        await Email.sendConfirmEmail(projectNames, requestedEmail, subscription.confirmKey);
```

Comment:
It might be good to add a try/catch block around all db requests, unless you have another error catching technique.